### ISSUE TITLE
Forecasting beyond 10 days

### ISSUE THREAD

**Issue by tvonich** [google-deepmind/graphcast#63]
There's seems to be some default setting that won't allow a forecast beyond 10 days. I've tried identifying where in the code that might be and have not been able to find the spot. 

Anyone successfully do this or know where to point me for fixing it?

**Comment by tewalds** [comment]
Take a look at `rollout.chunked_prediction`. It takes a `targets_template` that defines how many outputs to generate. If you give it a template with more outputs it'll generate more outputs. You can generate as many as you want (months?), but don't expect it to generate predictions that are good much beyond 10 days. At some point it'll generate worse predictions than climatology, but I'm not sure when that is. If you want to evaluate it you'll obviously also need to get real data (from ERA5 or HRES) as we haven't included more than 10 days of example data.

**Comment by tvonich** [comment]
> Hello, I'd like to ask how you made the prediction within ten days. I entered 20 time data, but starting from the third data, the prediction is compared with the original data. I want to predict the time after 20 data. What should I do? If you know, please tell me, thank you.

If you check out the #32 and #56 I expect this is your same issue. It has to do with where the inputs are extracted from the example data.

**Comment by ws1234321** [comment]
> > Hello, I'd like to ask how you made the prediction within ten days. I entered 20 time data, but starting from the third data, the prediction is compared with the original data. I want to predict the time after 20 data. What should I do? If you know, please tell me, thank you.
> 
> If you check out the #32 and #56 I expect this is your same issue. It has to do with where the inputs are extracted from the example data.

Thank you. I think I understand a little. I will make changes in an aligned way. Thank you very much for your advice.

**Comment by ws1234321** [comment]
> > Hello, I'd like to ask how you made the prediction within ten days. I entered 20 time data, but starting from the third data, the prediction is compared with the original data. I want to predict the time after 20 data. What should I do? If you know, please tell me, thank you.你好，我想问一下你是如何在十天内做出预测的。我输入了 20 个时间数据，但从第三个数据开始，将预测与原始数据进行比较。我想预测 20 个数据之后的时间。我该怎么办？如果你知道，请告诉我，谢谢。
> 
> If you check out the #32 and #56 I expect this is your same issue. It has to do with where the inputs are extracted from the example data. ❗ 🔄

Hi, I'm sorry to bother you again. I operate on the data in a left aligned and right aligned manner. I use data from 1 to 20, and the training and evaluation steps are both 4. Left aligned, the predicted data are 3, 4, 5, and 6. Right aligned, the predicted results are 17, 18, 19, and 20. If I want to predict 21, 22, 23, and 24, what should I do? Thank you in advance

**Comment by tewalds** [comment]
You need to generate a targets_template of the correct size. It can be filled with nans, but needs to be bigger than what you're currently passing in. You'll also need to generate the matching forcings.

**Comment by ws1234321** [comment]
> You need to generate a targets_template of the correct size. It can be filled with nans, but needs to be bigger than what you're currently passing in. You'll also need to generate the matching forcings.

Thank you very much for your help. I will make adjustments as you said and express my thanks again.

**Comment by tvonich** [comment]
> Take a look at `rollout.chunked_prediction`. It takes a `targets_template` that defines how many outputs to generate. If you give it a template with more outputs it'll generate more outputs. You can generate as many as you want (months?), but don't expect it to generate predictions that are good much beyond 10 days. At some point it'll generate worse predictions than climatology, but I'm not sure when that is. If you want to evaluate it you'll obviously also need to get real data (from ERA5 or HRES) as we haven't included more than 10 days of example data.

Timo, when you say it will generate worse error than climatology, does the error ever settle into a climatological plateau or does it grow forever?

**Comment by ElNino9495** [comment]
Hi, I have a doubt, if i have my input as 1/1/24 06:00  and 1/1/24 12:00  and my targets are from  1/1/24 18:00  to 1/2/24 12:00 (TARGET SIZE OF FOUR) and i have the forcings of the following, can i iteratively use the outputs and successfully predict so on for a year?

**Comment by tvonich** [comment]
Are you saying you want to do a one year forecast? That is possible, but
you need to create a target that is one year long. You can concatenate many
targets together to accomplish this. If there is no training, then it
doesn’t matter what they are as long as they are the right dimensions.

**Comment by ElNino9495** [comment]
Correct, but I'll also have to generate the corresponding forcings which will take a huge amount of time, right?

**Comment by tvonich** [comment]
No. It’s very quick. The model creates all necessary forcing during the
extract_input_target_forcing function. So if you specify a forecast of 40
steps (10 days), it makes all the needed forcings then.

Be aware that Graphcast has been shown to go unstable after 45 days-ish.

**Comment by ElNino9495** [comment]
I have managed to generate the forcings quickly but when i try to run my prediction model for 4 steps, the kernel always crashed. Is there something I can do to optimise my code?

**Comment by ElNino9495** [comment]
2024-09-23 11:07:11.281222: E external/xla/xla/service/slow_operation_alarm.cc:65] Constant folding an instruction is taking > 1s:

  %add.675 = s32[3114720]{0} add(s32[3114720]{0} %broadcast.2435, s32[3114720]{0} %constant.620), metadata={op_name="jit(<unnamed wrapped function>)/jit(main)/while/body/mesh2grid_gnn/_process/mesh2grid_gnn/_process_step/mesh2grid_gnn/add" source_file="/data4/home/rohitsuresh/graphcast/graphcast/typed_graph_net.py" source_line=142}
  
  this is the error i get

**Comment by tvonich** [comment]
Which grid and vertical resolution model are you using? What kind of GPU
are you trying to run it on?

**Comment by ElNino9495** [comment]
Its an nvidia a100 with 40gb ram and Im using the GraphCast_operational-ERA5-HRES_1979-2021-resolution_0.25-pressure_levels_13-mesh_2to6-precipitation_output_only.npz

**Comment by ElNino9495** [comment]
<img width="1182" alt="Screenshot 2024-09-24 at 7 13 21 AM" src="https://github.com/user-attachments/assets/ea47a585-1a49-4e04-a4a8-4d43f36e3d6d">
<img width="1151" alt="Screenshot 2024-09-24 at 7 13 31 AM" src="https://github.com/user-attachments/assets/ff376aeb-53b9-4c0e-969c-987851a6adef">
<img width="1024" alt="Screenshot 2024-09-24 at 7 13 46 AM" src="https://github.com/user-attachments/assets/b9f89aa0-8280-4687-801c-94e4b2840639">

This is what my inputs, targets and forcings look like before i convert them  to xarrays

**Comment by ElNino9495** [comment]
Hi, i was able to run it finally, but it took 20 minutes to make the predictions for 4 steps. Is this normal or can it be made faster?

**Comment by tvonich** [comment]
That's still very slow. With an a100 you should be able to do 4 steps in
30-40 seconds. Do you have the cuda-enabled version of jax in your
activated environment? I bet this is the problem and it is failing back to
CPU.

https://jax.readthedocs.io/en/latest/installation.html

**Comment by ElNino9495** [comment]
jax definitely uses my gpu. When i run gpustat, it shows that I'm currently
using the gpu.

**Comment by tvonich** [comment]
If that’s the case, then I do not know what the problem is. Maybe try a
verbose log file for more details. Good luck.
